linera-service/src/proxy/grpc.rs

```diff
@@ -468,6 +471,44 @@ where
         Ok(Response::new(certificate.try_into()?))
     }
 
+    #[instrument(skip_all, err(Display))]
+    async fn download_certificates(
+        &self,
+        request: Request<CertificatesBatchRequest>,
+    ) -> Result<Response<CertificatesBatchResponse>, Status> {
+        let hashes = request
+            .into_inner()
+            .hashes
+            .into_iter()
+            .map(linera_base::crypto::CryptoHash::try_from)
+            .collect::<Result<Vec<linera_base::crypto::CryptoHash>, _>>()?;
+
+        // Use 70% of the max message size as a buffer capacity.
+        // Leave 30% as overhead.
+        let mut grpc_message_limit: GrpcMessageLimiter<linera_chain::data_types::Certificate> =
+            GrpcMessageLimiter::new((GRPC_MAX_MESSAGE_SIZE as f64 * 0.7) as usize);
+
+        let mut certificates = vec![];
+
+        for certificate in self
+            .0
+            .storage
+            .read_certificates(hashes)
+            .await
+            .map_err(|err| Status::from_error(Box::new(err)))?
+        {
+            if grpc_message_limit.fits::<Certificate>(certificate.clone())? {
+                certificates.push(certificate);
+            } else {
+                break;
+            }
+        }
+
+        Ok(Response::new(CertificatesBatchResponse::try_from(
+            certificates,
+        )?))
```

Comment:
Also something to improve on in the future (not in this PR): We're converting the certificates twice now, once to check the encoded length, and then to actually encode them.

Reply:
I agree it's not ideal but not sure if it's possible - to encode (and know the size of the serialized element) protobuf consumes the variable. Unless you have some specific idea in mind.

Reply:
(No, I haven't looked into it in detail. Anyway, definitely not a blocker for this PR!)

Reply:
We could try to find a maximum possible size of the (`Executed`)`Block` and use that (there are a couple of collections inside the `Block` so it's not possible to determine "statically".

Reply:
But we do have a limit for that: https://github.com/linera-io/linera-protocol/pull/2454

Reply:
Right, it's configurable. In the PR I (implicitly) suggested 1 MG. :confused: 
But it can change per epoch, so we can't really use it here directly.